Add trimap to scanpy

eamid opened the pull request:
TriMap: Large-scale Dimensionality Reduction Using Triplets

TriMap is a dimensionality reduction method that uses triplet constraints to form a low-dimensional embedding of a set of points. The triplet constraints are of the form "point i is closer to point j than point k". The triplets are sampled from the high-dimensional representation of the points and a weighting scheme is used to reflect the importance of each triplet.

TriMap provides a significantly better global view of the data than the other dimensionality reduction methods such t-SNE, LargeVis, and UMAP. The global structure includes relative distances of the clusters, multiple scales in the data, and the existence of possible outliers. We define a global score to quantify the quality of an embedding in reflecting the global structure of the data. Also, TriMap is significantly faster than t-SNE, LargeVis, and UMAP and scales better to large datasets.

paper: https://arxiv.org/abs/1910.00204

code: https://github.com/eamid/trimap

eamid committed:
Merge pull request #1 from theislab/master

merge scanpy

eamid committed:
add trimap to scanpy

eamid committed:
add trimap to scanpy

fidelram commented:
Thanks for this contribution! Can you look at the travis errors? See https://github.com/theislab/scanpy/pull/797#issuecomment-536861482 for background.

To fix formatting errors you can use https://github.com/psf/black

Furthermore, can you add a small example to the docstring? (see for example: https://github.com/theislab/scanpy/blob/master/scanpy/external/tl/_palantir.py)

LuckyMD commented:
I guess this should probably be added to scanpy external as other external packages are?

fidelram commented:
@LuckyMD I agree.

eamid committed:
moved trimap to external packages

eamid committed:
Merge pull request #2 from theislab/master

Mege master with trimap

eamid commented:
Thanks for getting back to me. I fixed the formatting errors and moved trimap to scanpy external. 
trimap is no longer imported by default, so the overall import time is unaffected. 
I also added an example to the docstring.

Please let me know if further fixes are required.

eamid committed:
one line summary

eamid committed:
Merge branch 'master' into master

eamid committed:
one line summary

eamid committed:
Merge branch 'master' of https://github.com/eamid/scanpy

eamid committed:
one line summary

eamid committed:
one line summary

eamid committed:
one line summary

eamid commented:
I fixed the merge conflicts. Hope it is all good now!
Please let me know if you need more changes.

flying-sheep commented on scanpy/external/exporting.py:
```diff
@@ -450,7 +450,7 @@ def cellbrowser(
         By default (or when specifying `'all'` or `None`), these keys are tried:
         [`"tsne"`, `"umap"`, `"pagaFa"`, `"pagaFr"`, `"pagaUmap"`, `"phate"`,
-        `"fa"`, `"fr"`, `"kk"`, `"drl"`, `"rt"`].
+        `"fa"`, `"fr"`, `"kk"`, `"drl"`, `"rt"`, `"trimap"`].
```
Nice catch. @theislab/scanpy: we should create this string using code, not by requiring people to find and update it.

flying-sheep commented on scanpy/external/tl/_trimap.py:
```diff
@@ -0,0 +1,148 @@
+    init_pos=None,
+    verbose=False,
+    copy=False,
+):
```
Please add type hints here. Check other methods for how to do `metric` using `compat.Literal`.

Don’t forget the return type!

flying-sheep commented on scanpy/external/tl/_trimap.py:
```diff
@@ -0,0 +1,148 @@
+    }
+
+    Input
+    ------
```
Please call this section `Parameters`, use numpydoc style and get rid of “default” (it’s added automatically)

flying-sheep commented on scanpy/external/tl/_trimap.py:
```diff
@@ -0,0 +1,148 @@
+    tuple (knn_nbrs, knn_distances) (default = None)
+
+    apply_pca: Apply PCA to reduce the dimensions to 100 if necessary before the
+    nearest-neighbor calculation (default = True)
```
I think the default should be `None`, which means “True, but throw a warning if we have to run sc.pp.pca”. Also the parameter seems to be missing?

flying-sheep commented on scanpy/external/tl/_trimap.py:
```diff
@@ -0,0 +1,148 @@
+    nearest-neighbor calculation (default = True)
+
+    opt_method: Optimization method ('sd': steepest descent,  'momentum': GD
+    with momentum, 'dbd': GD with momentum delta-bar-delta (default))
```
I don’t see this in the actual parameters, where is it?

Please sort the parameters like they appear in the function and make sure every documented one is also present and vice versa.

flying-sheep commented on scanpy/external/tl/_trimap.py:
```diff
@@ -0,0 +1,148 @@
+    opt_method: Optimization method ('sd': steepest descent,  'momentum': GD
+    with momentum, 'dbd': GD with momentum delta-bar-delta (default))
+
+    verbose: Print the progress report (default = True)
```
Please derive this from `sc.settings.verbosity` by default. (`None` by default)

flying-sheep commented on scanpy/external/tl/_trimap.py:
```diff
@@ -0,0 +1,148 @@
+    (default = 500.0)
+
+    return_seq: Return the sequence of maps recorded every 10 iterations
+    (default = False)
```
This parameter is also absent. If you conditionally want to return a non-AnnData, please replace “copy” with “inplace” (the semantics are “inplace=True” → modify “adata”, “inplace=False” → return a different data structure)

flying-sheep commented on scanpy/external/tl/_trimap.py:
```diff
@@ -0,0 +1,148 @@
+
+def trimap(
+    adata,
+    n_dims=2,
```
Please add a `*,` after `adata,` or `n_dims: int = 2,`.

This way we’re free to reorganize subsequent parameters without it being a breaking change.

flying-sheep commented on scanpy/external/tl/_trimap.py:
```diff
@@ -0,0 +1,148 @@
+        raise ValueError(
+            'metric \'' + metric + '\' is currently not supported '
+            'by trimap. acceptable values are \'euclidean\', \'manhattan\', '
+            '\'angular\', and \'hamming\'.'
```
please avoid unnecessary escaping and prefer f-strings to concatenating strings with `+`:

```py
f'Metric {metric!r} is currently not supported by trimap. '
'Acceptable values are {", ".join(map(repr, allowed_metrics))}'
```

flying-sheep commented on scanpy/external/tl/_trimap.py:
```diff
@@ -0,0 +1,148 @@
+    logg.info(
+        '    finished',
+        time=start,
+        deep=('added\n' "    'X_trimap', TriMap coordinates (adata.obsm)"),
```
```suggestion
        deep="added\n    'X_trimap', TriMap coordinates (adata.obsm)",
```

flying-sheep reviewed (changes requested):
Looks good apart from some details, and I liked the method when it saw it being presented a few months ago.

eamid committed:
type hints and doc

eamid committed:
blacked

eamid committed:
blacked

eamid commented:
Thanks @flying-sheep for the thorough feedback! I made the changes. There is still a Travis CI error about slow_to_import modules. Since trimap is now in external, I am now sure how this test is being affected.

flying-sheep commented:
Very weird. Importing `scipy.sparse` shouldn’t import `scipy.stats`, and you didn’t add any other imports. Could be that this is a change in some updated version of something.

Does the test fail for you? Then you could find out how it gets imported using `import-profiler`

eamid commented:
Thanks @flying-sheep! I haven't added any new imports other those for type hints and copying anndata: [https://github.com/eamid/scanpy/compare/9ae6c19...b7ed705](https://github.com/eamid/scanpy/compare/9ae6c19...b7ed705)

My previous commit passed the tests. Could this be because of recent commits on your side?

gokceneraslan commented:
I have the same issue here https://github.com/theislab/scanpy/pull/945, so it's not specific to this PR,fyi.

flying-sheep commented:
To help debug it, can one of you do what I suggested? I don’t have the problem locally and if one of you does see it, it would be much easier to find out which dependency started to import that module.

flying-sheep reviewed (changes requested):
The PR looks really clean now, congrats!

One thing left I just realized: the bibtex code for the reference will look pitiful when rendered as reStructuredText. Can you please add yourself to `references.rst` and just cite yourself as `[Amid19]_` in the docstring instead of using that “Reference” section?

eamid committed:
reference

eamid commented:
Thank you @flying-sheep! All done.

eamid commented:
@flying-sheep just following up on the PR. Thanks!

flying-sheep commented on scanpy/external/tl/_trimap.py:
```diff
@@ -0,0 +1,129 @@
+    -------
+    Depending on `copy`, returns or updates `adata` with the following fields.
+
+    **X_trimap** : `np.ndarray`, (`adata.obs`, shape=[n_samples, n_components], dtype `float`)
```
```suggestion
    **X_trimap** : :class:`~numpy.ndarray`, (:attr:`~anndata.AnnData.obsm`, shape=(n_samples, n_components), dtype `float`)
```

One typo I missed: `.obs` → `.obsm`.

eamid committed:
Update scanpy/external/tl/_trimap.py typo

Co-Authored-By: Philipp A. <flying-sheep@web.de>

flying-sheep commented:
Great, thank you!

eamid commented:
Thank you @flying-sheep!
